### PR TITLE
Canonical query examples

### PR DESCRIPTION
**Pull request by kara**
This PR adds canonical examples for view and content queries. 

cc @jelbourn

### CODE REVIEW

File: packages/core/src/render3/query.ts [hunk 1]
```diff
@@ -394,3 +394,38 @@ export function queryRefresh(queryList: QueryList<any>): boolean {
...
+ * is retrieved from data[directiveIndex][dataIndex].
+ *
+ * If a non-numerical value is passed as the second arg, this is a "set" request and
+ * the QueryList will be saved in directive memory at the data index indicated.
+ *
+ * @param dataIndex The index of the data in directive memory array
+ * @param directiveIndexOrValue The directive index to retrieve or the value to set
+ * @returns Either the instance or a QueryList, depending on dataIndex
+ */
+export function directiveMemory<T>(
```
Comment (mhevery) [inline packages/core/src/render3/query.ts]: I think of `directiveMemory` as general function for staring data, not specific to queries. 
- It should return `T` not `QueryList<T>`
- It should not be in `query.ts` but rather in `instructions.ts` as a generic storage mechanism.

File: packages/core/src/render3/instructions.ts [hunk 2]
```diff
@@ -1858,6 +1860,16 @@ export function getRenderer(): Renderer3 {
   return renderer;
 }
 
+export function getData(): any {
```
Comment (mhevery) [inline packages/core/src/render3/instructions.ts]: you don't need `getData` if `directiveMemory` is moved here.

File: packages/core/src/render3/query.ts [hunk 3]
```diff
@@ -394,3 +394,38 @@ export function queryRefresh(queryList: QueryList<any>): boolean {
...
+    dataIndex: number, directiveIndexOrValue: number | QueryList<T>): QueryList<T>|T {
+  const data = getData();
+  if (typeof directiveIndexOrValue === 'number') {
+    ngDevMode && assertDataInRange(directiveIndexOrValue as number, data);
+    ngDevMode && assertEqual(
+                     Array.isArray(data[directiveIndexOrValue]), true,
+                     'expected data[directiveIndexOrValue] to be an array');
+    return data[directiveIndexOrValue][dataIndex as number];
+  } else {
```
Comment (mhevery) [inline packages/core/src/render3/query.ts]: `else` branch is not needed per our discussion

File: packages/core/test/render3/compiler_canonical_spec.ts [hunk 4]
```diff
@@ -238,6 +238,137 @@ describe('compiler specification', () => {
...
+        class ContentQueryComponent {
+          @ContentChild(SomeDirective) someDir: SomeDirective;
+
+          // NORMATIVE
+          static ngComponentDef = r3.defineComponent({
+            type: ContentQueryComponent,
+            tag: 'content-query-component',
+            factory: function ContentQueryComponent_Factory() {
+              r3.dM(1, r3.Q(SomeDirective, false));
+              return new ContentQueryComponent();
```
Comment (mhevery) [inline packages/core/test/render3/compiler_canonical_spec.ts]: `return [new ContentQueryComponent(), r3.Q(null, SomeDirective, false))`

File: packages/core/test/render3/compiler_canonical_spec.ts [hunk 5]
```diff
@@ -238,6 +238,137 @@ describe('compiler specification', () => {
...
+
+          // NORMATIVE
+          static ngComponentDef = r3.defineComponent({
+            type: ViewQueryComponent,
+            tag: 'view-query-component',
+            factory: function ViewQueryComponent_Factory() { return new ViewQueryComponent(); },
+            template: function ViewQueryComponent_Template(ctx: ViewQueryComponent, cm: boolean) {
+              let tmp: any;
+              if (cm) {
+                r3.m(0, r3.Q(SomeDirective, false));
```
Comment (mhevery) [inline packages/core/test/render3/compiler_canonical_spec.ts]: could we simplify to: `r3.Q(0, SomeDirective, false);`

**Review by mhevery — CHANGES_REQUESTED** [review]

File: packages/core/src/render3/query.ts [hunk 6]
```diff
@@ -371,12 +371,16 @@ export type QueryList<T> = viewEngine_QueryList<T>;
 export const QueryList: typeof viewEngine_QueryList = QueryList_ as any;
 
 export function query<T>(
```
Comment (mhevery) [inline packages/core/src/render3/query.ts]: Since `query` is public can we add jsdoc?

**Review by mhevery — APPROVED** [review]
Please add JSDoc, LGTM otherwise

**Review by mhevery — APPROVED** [review]

File: packages/core/src/render3/query.ts [hunk 7]
```diff
@@ -370,6 +370,16 @@ class QueryList_<T>/* implements viewEngine_QueryList<T> */ {
 export type QueryList<T> = viewEngine_QueryList<T>;
 export const QueryList: typeof viewEngine_QueryList = QueryList_ as any;
 
+/**
+ * Creates and returns a QueryList,
+ *
+ * @param memoryIndex The index in memory where the QueryList should be saved. If null,
+ * this is is a content query and the QueryList will be saved later through directiveCreate.
```
Comment (JoostK) [inline packages/core/src/render3/query.ts]: "is is" -> "is"

**Pull request closed without merging**